syntax = "proto3";

package university;

option py_generic_services = true;

// Enum for loan status
enum LoanStatus {
  ONGOING = 0;
  RETURNED = 1;
  LATE = 2;
}

// Book messages
message Book {
  string id = 1;
  string title = 2;
  string author = 3;
  string isbn = 4;
  string publisher = 5;
  int32 page_count = 6;
  int32 stock = 7;
}

message GetBookRequest {
  string id = 1;
}

message GetBookResponse {
  Book book = 1;
}

message ListBooksRequest {
  int32 page_size = 1;
  int32 page_number = 2;
}

message ListBooksResponse {
  repeated Book books = 1;
  int32 total_count = 2;
}

message CreateBookRequest {
  Book book = 1;
}

message CreateBookResponse {
  Book book = 1;
}

message UpdateBookRequest {
  Book book = 1;
}

message UpdateBookResponse {
  Book book = 1;
}

message DeleteBookRequest {
  string id = 1;
}

message DeleteBookResponse {
  bool success = 1;
  string message = 2;
}

// Student messages
message Student {
  string id = 1;
  string name = 2;
  string student_number = 3;
  string email = 4;
  bool is_active = 5;
}

message GetStudentRequest {
  string id = 1;
}

message GetStudentResponse {
  Student student = 1;
}

message ListStudentsRequest {
  int32 page_size = 1;
  int32 page_number = 2;
}

message ListStudentsResponse {
  repeated Student students = 1;
  int32 total_count = 2;
}

message CreateStudentRequest {
  Student student = 1;
}

message CreateStudentResponse {
  Student student = 1;
}

message UpdateStudentRequest {
  Student student = 1;
}

message UpdateStudentResponse {
  Student student = 1;
}

message DeleteStudentRequest {
  string id = 1;
}

message DeleteStudentResponse {
  bool success = 1;
  string message = 2;
}

// Loan messages
message Loan {
  string id = 1;
  string student_id = 2;
  string book_id = 3;
  string loan_date = 4;
  string return_date = 5;
  LoanStatus status = 6;
}

message GetLoanRequest {
  string id = 1;
}

message GetLoanResponse {
  Loan loan = 1;
}

message ListLoansRequest {
  int32 page_size = 1;
  int32 page_number = 2;
}

message ListLoansResponse {
  repeated Loan loans = 1;
  int32 total_count = 2;
}

message CreateLoanRequest {
  string student_id = 1;
  string book_id = 2;
}

message CreateLoanResponse {
  Loan loan = 1;
}

message ReturnBookRequest {
  string loan_id = 1;
}

message ReturnBookResponse {
  Loan loan = 1;
}

// Services
service BookService {
  rpc GetBook(GetBookRequest) returns (GetBookResponse);
  rpc ListBooks(ListBooksRequest) returns (ListBooksResponse);
  rpc CreateBook(CreateBookRequest) returns (CreateBookResponse);
  rpc UpdateBook(UpdateBookRequest) returns (UpdateBookResponse);
  rpc DeleteBook(DeleteBookRequest) returns (DeleteBookResponse);
}

service StudentService {
  rpc GetStudent(GetStudentRequest) returns (GetStudentResponse);
  rpc ListStudents(ListStudentsRequest) returns (ListStudentsResponse);
  rpc CreateStudent(CreateStudentRequest) returns (CreateStudentResponse);
  rpc UpdateStudent(UpdateStudentRequest) returns (UpdateStudentResponse);
  rpc DeleteStudent(DeleteStudentRequest) returns (DeleteStudentResponse);
}

service LoanService {
  rpc GetLoan(GetLoanRequest) returns (GetLoanResponse);
  rpc ListLoans(ListLoansRequest) returns (ListLoansResponse);
  rpc CreateLoan(CreateLoanRequest) returns (CreateLoanResponse);
  rpc ReturnBook(ReturnBookRequest) returns (ReturnBookResponse);
}